swarm/src/handler.rs

```diff
@@ -224,6 +275,76 @@ pub trait ConnectionHandler: Send + 'static {
     {
         ConnectionHandlerSelect::new(self, other)
     }
+
+    /// informs the handler about an event coming from the outside in the handler.
+    fn on_behaviour_event(&mut self, _event: Self::InEvent) {}
+
+    fn on_event(
+        &mut self,
+        _event: StreamEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+    }
+}
+
+/// Enumeration with the list of the possible stream events
+/// to pass to [`on_event`](ConnectionHandler::on_event).
+pub enum StreamEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IOI, OOI> {
```

Comment:
An alternative to explore (not saying it is a good idea) would be to take a single trait parameter, namely `ConnectionHandler` here.

```suggestion
pub enum StreamEvent<'a, Handler: ConnectionHandler> {
```

Reply:
Thanks for bringing this Max,  
yeah it works but we need to relax the sizedness to `Handler: ConnectionHandler + ?Sized` because `on_event`'s argument `event: StreamEvent<Self>`  which afaik will probably be problematic if we in the future define methods on `ConnectionHandler` that take `self` and `StreamEvent<Self>` cause `Self` must be `Sized` and `StreamEvent<Self>` may be `?Sized`. Would you prefer that over the multiple generic parameters?

Reply:
> yeah it works but we need to relax the sizedness to `Handler: ConnectionHandler + ?Sized` because `on_event`'s argument `event: StreamEvent<Self>` which afaik will probably be problematic if we in the future define methods on `ConnectionHandler` that take `self` and `StreamEvent<Self>` cause `Self` must be `Sized` and `StreamEvent<Self>` may be `?Sized`.

I am not sure I am following. In which case would `ConnectionHandler` not be `Sized`? In which case would this be problematic?

Reply:
The approach you suggest on https://github.com/libp2p/rust-libp2p/pull/3085#discussion_r1016988595:
```suggestion
pub enum StreamEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IOI, OOI> {
```
implies relaxing `StreamEvent<Handler>` to `StreamEvent<Handler: ConnectionHandler + ?Sized>`.